bolt-boost/src/server.rs

```diff
@@ -0,0 +1,588 @@
+
+// lazy_static! {
+//     pub static ref CHECK_RECEIVED_COUNTER: IntCounter =
+//         IntCounter::new("checks", "successful /check requests received").unwrap();
+// }
+
+// Extra state available at runtime
+#[derive(Clone)]
+pub struct BuilderState {
+    #[allow(unused)]
+    config: Config,
+    constraints: ConstraintsCache,
+}
+
+impl BuilderApiState for BuilderState {}
+
+impl BuilderState {
+    pub fn from_config(config: Config) -> Self {
+        Self {
+            config,
+            constraints: ConstraintsCache::new(),
+        }
+    }
+}
+
+/// An extended builder-API that implements the constraints-API as defined in
+/// the spec: <https://chainbound.github.io/bolt-docs/api/builder>.
+///
+/// The added endpoints are defined in [extra_routes](ConstraintsApi::extra_routes).
+pub struct ConstraintsApi;
+
+#[async_trait]
+impl BuilderApi<BuilderState> for ConstraintsApi {
+    /// Register a validator with the builder.
+    ///
+    /// We intercept this call since it happens periodically and we use it to clean
+    /// up old constraints.
+    async fn register_validator(
+        registrations: Vec<ValidatorRegistration>,
+        req_headers: HeaderMap,
+        state: PbsState<BuilderState>,
+    ) -> eyre::Result<()> {
+        let (slot, _) = state.get_slot_and_uuid();
+
+        info!("Cleaning up constraints before slot {slot}");
+        state.data.constraints.remove_before(slot);
+
+        register_validator(registrations, req_headers, state).await
+    }
+
+    /// Gets the extra routes for supporting the constraints API as defined in
+    /// the spec: <https://chainbound.github.io/bolt-docs/api/builder>.
+    fn extra_routes() -> Option<Router<PbsState<BuilderState>>> {
+        let mut router = Router::new();
+        router = router.route(SUBMIT_CONSTRAINTS_PATH, post(submit_constraints));
+        router = router.route(DELEGATE_PATH, post(delegate));
+        router = router.route(REVOKE_PATH, post(revoke));
+        router = router.route(GET_HEADER_WITH_PROOFS_PATH, get(get_header_with_proofs));
+        Some(router)
+    }
+}
+
+/// Submit signed constraints to the builder.
+/// Spec: <https://chainbound.github.io/bolt-docs/api/builder#constraints>
+#[tracing::instrument(skip_all)]
+async fn submit_constraints(
+    State(state): State<PbsState<BuilderState>>,
+    Json(constraints): Json<Vec<SignedConstraints>>,
+) -> Result<impl IntoResponse, PbsClientError> {
+    info!("Submitting {} constraints to relays", constraints.len());
+    // Save constraints for the slot to verify proofs against later.
+    for signed_constraints in &constraints {
+        if let Err(e) = state.data.constraints.insert(
+            signed_constraints.message.slot,
+            signed_constraints.message.clone(),
+        ) {
+            error!(slot = signed_constraints.message.slot, error = %e, "Failed to save constraints");
+            return Err(PbsClientError::BadRequest);
+        }
+    }
```

Comment:
Related to the comment above on clearing the cache upon registering validators, I think this implementation may introduce some OOM risks:
- the `slot` is not validated, that is we don't check whether as a proposer I'm actually going to propose a block at that slot
- in the underlying hashmap of the cache, every `Vec` value can grow indefinitely.
As such I'm not sure how this endpoint is resistant to spamming. I'd propose an LRU cache or FIFOMap with a certain capacity, and bound checks on vectors not growing past `MAX_CONSTRAINTS_PER_SLOT`

Reply:
Added some simple safeguards to minimize OOM risk